hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/keyvalue/KvMetadataAvroResolver.java

```diff
@@ -93,90 +125,125 @@ public KvMetadata resolveMetadata(
-                                   .unionOf()
-                                   .nullType()
-                                   .and().booleanType()
-                                   .and().intType()
-                                   .and().longType()
-                                   .and().floatType()
-                                   .and().doubleType()
-                                   .and().stringType()
-                                   .endUnion()
-                                   .nullDefault();
+                    schema = schema.name(path).type(Schemas.OBJECT_SCHEMA).withDefault(null);
                     break;
                 default:
-                    throw new IllegalArgumentException("Unknown type: " + type.getTypeFamily());
+                    throw new IllegalArgumentException("Unknown type: " + field.getType().getTypeFamily());
             }
         }
         return schema.endRecord();
     }
 
-    private QueryPath[] paths(List<TableField> fields) {
-        return fields.stream().map(field -> ((MapTableField) field).getPath()).toArray(QueryPath[]::new);
+    private void validate(Schema schema, Map<QueryPath, MappingField> fieldsByPath) {
+        if (schema.getType() != Schema.Type.RECORD) {
+            throw new IllegalArgumentException("Schema must be an Avro record");
+        }
+        for (Entry<QueryPath, MappingField> entry : fieldsByPath.entrySet()) {
+            String path = entry.getKey().getPath();
+            if (path == null) {
+                continue;
+            }
+            Schema.Field schemaField = schema.getField(path);
+            if (schemaField == null) {
+                throw new IllegalArgumentException("Field '" + path + "' does not exist in schema");
+            }
+            Schema.Type schemaFieldType = unwrapNullableType(schemaField.schema()).getType();
+            QueryDataTypeFamily mappingFieldType = entry.getValue().type().getTypeFamily();
+            VarargsConsumer<Schema.Type> assertCompatible = types -> {
+                if (!asList(types).contains(schemaFieldType)) {
+                    throw new IllegalArgumentException(schemaFieldType + " schema type is incompatible with "
+                            + mappingFieldType + " mapping type");
+                }
+            };
+            // Logical types are not supported, which require changes in IntegerConverter and LongConverter.
+            switch (mappingFieldType) {
+                case BOOLEAN:
+                    assertCompatible.accept(Schema.Type.BOOLEAN, Schema.Type.STRING);
+                    break;
+                case TINYINT:
+                case SMALLINT:
+                case INTEGER:
+                case BIGINT:
+                case REAL:
+                case DOUBLE:
+                case DECIMAL:
+                    assertCompatible.accept(Schema.Type.INT, Schema.Type.LONG, Schema.Type.FLOAT, Schema.Type.DOUBLE,
```

Comment:
are those the same conditions as in `AvroUpsertTarget.CONVERSION_PREFS`? If so, can we do that without repetition?